src/GraphQLParser/AST/GraphQLDescription.cs

```diff
@@ -0,0 +1,9 @@
+namespace GraphQLParser.AST
+{
+    public class GraphQLDescription : ASTNode
```

Comment:
@Shane32  Where is `Location` for `GraphQLDescription` syntax node ?

Reply:
I think I didn't put one in because the syntax for the parent node technically includes the description, and so the start of the description is the same as the start of the type definition node.  At least that's maybe why I did it???  It's been 4 months so I can't remember my reasoning exactly.

Maybe it should be in there anyway though, so that there's a way to determine the end of the description node?  I guess that would be a new class called `GraphQLDescriptionFull` ... ??

Reply:
> the syntax for the parent node technically includes the description, and so the start of the description is the same as the start of the type definition node. 

Nope. Those are different syntax nodes with different locations. `GraphQLComment` node has `Location` for example.

> I guess that would be a new class called `GraphQLDescriptionFull`

Yep. Maybe new, maybe existing. I should remember why I did only one `GraphQLComment` class without `GraphQLCommentFull` counterpart. I have already started review, so I can do all the necessary edits myself.

Reply:
![image](https://user-images.githubusercontent.com/6377684/140585757-47185151-851c-4260-b42f-b71d3d336168.png)

Reply:
Pretty sure the library is already written this way -- to match the spec.  So a type definition node will have its Location start and end points exactly as described by the spec and shown above.

I don't see any reason why a description node can't have a Location also

Reply:
@Shane32 I burst between two options - do all the necessary edits and improvements or only part. In the first case there will be backward incompatible changes: https://github.com/graphql-dotnet/parser/issues/141. Not so serious, but still backward incompatible. In the second case, compatibility will preserved, but the code will be far from perfect. Also I noticed that you missed description for `GraphQLSchemaDefinition`. I want to remove `GraphQLTypeDefinitionWithDescription` class, rename `IHasDescription` and make `IgnoreOptions` as flags enum:
```
None = 0,
Locations = 1,
Comments = 2,
All = Locations | Comments,
```

Personally, I would prefer to release v8 and continue to working on GraphQL v5.

Reply:
> Pretty sure the library is already written this way -- to match the spec. So a type definition node will have its Location start and end points exactly as described by the spec and shown above.

I have not yet thought about it.

> I don't see any reason why a description node can't have a Location also

Yes, I will add a property.

Reply:
> I want to remove GraphQLTypeDefinitionWithDescription class

Why?

Reply:
This is not a real class. It only creates an additional level in the inheritance chain. In fact, the property could be added to the basic class `GraphQLTypeDefinition`. Also `GraphQLSchemaDefinition` class **should** have this property but schema definition does not inherit from `GraphQLTypeDefinition`. It would be strange to make one more `WithDescription` base class. Only `GraphQLTypeExtensionDefinition` has no description due to the spec. Nothing terrible is that this property will appear in this class. We do not set it.

Reply:
It allocates extra memory.  Everything we have done in GraphQL.NET and the parser library has all been with a goal of creating no extra allocations.  Why if we are not going to carry through?

If nothing else, it should be set up with a property with no backing field, and then change the 'withdescription' version as an internal class that has a backing field for the property, as is already done for comments.

Reply:
It allocates only for `GraphQLTypeExtensionDefinition`.

> and then change the 'withdescription' version as an internal class that has a backing field for the property, as is already done for comments.

I thought about it. At first I wanted to go this way. Then I refused it - combinations appear. Too difficult to read/maintain.

No big deal especially since only server parts of the schema may have descriptions and parsing in ordinary scenarios is a one-time operation.

Reply:
There is no added complexity. It is not complex. It is not difficult to write. In fact you simply don’t need to remove optimized functionality that already exists in the library. I do not see any reason to take something optimized and unoptimize it. It does not make sense.

Reply:
The current implementation adds a couple lines of code to the entire library. As in like simply 2-3 lines of code. That is not complex. It is not harder to read. There is no reason to deoptimize the library for the sake of 2 lines of code.

Reply:
I will think. There is a small chance that schema extensions can also have descriptions (although it is unlikely) - https://github.com/graphql/graphql-spec/issues/900

Reply:
Well, everything turned out to be simple and difficult at the same time. See http://spec.graphql.org/October2021/#sec-Type-Extensions or even more old http://spec.graphql.org/June2018/#sec-Type-Extensions :
![изображение](https://user-images.githubusercontent.com/21261007/140623605-8d0d9a92-e243-4dab-a8f2-41269c10b9cb.png)

Now parser supports (partially) only ObjectTypeExtension. Wow! :) No support for other extensions at all, no tests. The good news is that Description property will not exist on GraphQLTypeExtensionDefinition when it stops inherit from `GraphQLTypeDefinition`. All extensions **should** have a separate new base class. Also we should add a bunch of new classes to cover each kind of extension. Looks like a good feature for v8. So I suggest to merge this as is (after your review of course) and continue to work on separate issue - https://github.com/graphql-dotnet/parser/issues/143.

Reply:
#154